components/suggestion/switch_channel_provider.jsx

```diff
@@ -584,16 +596,39 @@ export default class SwitchChannelProvider extends Provider {
         };
     }
 
-    formatUnreadChannelsAndDispatch(resultsCallback) {
-        const getChannel = makeGetChannel();
-
-        const unreadChannelIds = getSortedUnreadChannelIds(getState(), false);
+    fetchAndFormatRecentChannels(resultsCallback) {
+        const state = getState();
+        const recentChannels = getAllRecentChannels(state);
+        const channels = this.wrapChannels(recentChannels, Constants.MENTION_RECENT_CHANNELS);
+        if (channels.length === 0) {
+            prefix = '';
+            this.startNewRequest('');
+            this.fetchChannels(resultsCallback);
+        }
+        const sortedChannels = channels.sort(sortChannelsByRecencyAndTypeAndDisplayName).slice(0, 20);
+        const channelNames = sortedChannels.map((wrappedChannel) => wrappedChannel.channel.id);
+        resultsCallback({
+            matchedPretext: '',
+            terms: channelNames,
+            items: sortedChannels,
+            component: ConnectedSwitchChannelSuggestion,
+        });
+    }
 
-        const channels = [];
-        for (let i = 0; i < unreadChannelIds.length; i++) {
-            const channel = getChannel(getState(), {id: unreadChannelIds[i]}) || {};
+    wrapChannels(channels, channelType) {
+        const currentChannel = getCurrentChannel(getState());
+        const members = getMyChannelMemberships(getState());
 
+        const channelList = [];
+        for (let i = 0; i < channels.length; i++) {
+            const channel = channels[i];
+            if (channel.id === currentChannel.id) {
+                continue;
+            }
             let wrappedChannel = {channel, name: channel.name, deactivated: false};
+            if (members[channel.id]) {
+                wrappedChannel.last_viewed_at = members[channel.id].last_viewed_at;
```

Comment:
Have done the suggested changes I think we should be good now.

Reply:
@hmhealey this will not work here as myPreferences only have values shown in screenshot
<img width="825" alt="Screenshot 2021-06-08 at 1 10 09 AM" src="https://user-images.githubusercontent.com/16203333/121078278-5a8cef00-c7f6-11eb-9291-3e4e532565b7.png">

Reply:
@hmhealey Any suggestion for this one, this is the blocker I guess now.

Reply:
That's the only channel_approximate_view_time preference in there? As far as I know, there should be one for each channel, but the logic around it is messy, so I wouldn't be surprised if there's something I've missed.

Looking at that logic again though, unfortunately we only update that preference once every 3 hours ([link](https://github.com/mattermost/mattermost-webapp/blob/master/packages/mattermost-redux/src/actions/channels.ts#L788)), so it's not fully suitable for this. I think we're going to have to fix the behaviour of last_viewed_at to make this actually work the way we want in all cases.

I think we can continue the review process on this PR for the time being, but we'll have to hold off on merging it until we figure out what to do with that. As it is, getting a proper sense of the last viewed channel will require some bigger changes that need to be investigated further.

Reply:
@hmhealey Okay I will pause the changes related to this and finish up other suggestions.